seccomp_test.go

```diff
@@ -701,3 +675,265 @@ func TestCreateActKillProcessFilter(t *testing.T) {
+			t.Skipf("Skipping test: %s", err)
+		}
+
+		t.Errorf("Error getting API level: %s", err)
+	} else {
+		t.Logf("Got API level %v", api)
+		if api < 5 {
+			err = SetAPI(5)
+			if err != nil {
+				t.Skipf("Skipping test: API level %d is less than 5 and could not set it to 5", api)
+				return
+			}
+		}
+	}
+
+	arch, err := GetNativeArch()
+	if err != nil {
+		t.Errorf("Error in GetNativeArch(): %s", err)
+		return
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Errorf("Error in Getwd(): %s", err)
+		return
+	}
+
+	filter, err := NewFilter(ActAllow)
+	if err != nil {
+		t.Errorf("Error creating filter: %s", err)
+	}
+	defer filter.Release()
+
+	// Seccomp notification is only supported on single-thread filters
+	err = filter.SetTsync(false)
+	if err != nil {
+		t.Errorf("Error setting tsync on filter: %s", err)
+	}
+
+	call, err := GetSyscallFromName("chdir")
+	if err != nil {
+		t.Errorf("Error getting syscall number: %s", err)
+	}
+
+	err = filter.AddRule(call, ActNotify)
+	if err != nil {
+		t.Errorf("Error adding rule to log syscall: %s", err)
+	}
+
+	nonExistentPath, err := syscall.BytePtrFromString("/non-existent-path")
+	if err != nil {
+		t.Errorf("Error converting string: %s", err)
+	}
+	currentWorkingDirectory, err := syscall.BytePtrFromString(cwd)
+	if err != nil {
+		t.Errorf("Error converting string: %s", err)
+	}
+
+	tests := []notifTest{
+		{
+			syscall:     call,
+			args:        [6]uintptr{uintptr(unsafe.Pointer(nonExistentPath)), 0, 0, 0, 0, 0},
+			arch:        arch,
+			respErr:     0,
+			respVal:     0,
+			respFlags:   NotifRespFlagContinue,
+			expectedErr: syscall.ENOENT,
+			expectedVal: ^uint64(0), // -1
+		},
+		{
+			syscall:     call,
+			args:        [6]uintptr{uintptr(unsafe.Pointer(currentWorkingDirectory)), 0, 0, 0, 0, 0},
+			arch:        arch,
+			respErr:     0,
+			respVal:     0,
+			respFlags:   NotifRespFlagContinue,
+			expectedErr: syscall.Errno(0),
+			expectedVal: 0,
+		},
+		{
+			syscall:     call,
+			args:        [6]uintptr{uintptr(unsafe.Pointer(nonExistentPath)), 0, 0, 0, 0, 0},
+			arch:        arch,
+			respErr:     int32(syscall.ENOMEDIUM),
+			respVal:     ^uint64(0), // -1
+			respFlags:   0,
+			expectedErr: syscall.ENOMEDIUM,
+			expectedVal: ^uint64(0), // -1
+		},
+		{
+			syscall:     call,
+			args:        [6]uintptr{uintptr(unsafe.Pointer(currentWorkingDirectory)), 0, 0, 0, 0, 0},
+			arch:        arch,
+			respErr:     int32(syscall.EPIPE),
+			respVal:     ^uint64(0), // -1
+			respFlags:   0,
+			expectedErr: syscall.EPIPE,
+			expectedVal: ^uint64(0), // -1
+		},
+	}
+
+	seccompFdChan := make(chan ScmpFd)
+	errorChan := make(chan error, 2)
+	infoChan := make(chan string)
+	done := make(chan struct{})
+
+	go func() {
+		// Lock this goroutine to it's current kernel thread; otherwise the go runtime may
+		// switch us to a different OS thread, bypassing the seccomp notification filter.
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
```

Comment:
Documentation of `UnlockOSThread` says:
> // Before calling UnlockOSThread, the caller must ensure that the OS
// thread is suitable for running other goroutines. If the caller made
// any permanent changes to the state of the thread that would affect
// other goroutines, it should not call this function and thus leave
// the goroutine locked to the OS thread until the goroutine (and
// hence the thread) exits.

Since we don't _unload_ this filter (because it's not possible), I assume it would be correct to not call UnlockOSThread() here. Otherwise it could happen that by random choice this thread is re-used. In this test this seems improable but if it's true then it would be good to give the right example here. I'm not an expert and just reasoning from the docs, I have no proof of this.

Reply:
Thanks, that's a good catch. I had missed that description on UnlockOSThread(), and I agree with your reasoning that we should not unlock the goroutine from the OS thread given that the seccomp filter is already acting on it.

Reply:
@yvesf Thanks for testing. I have the same error when trying `-count 2` like you did. I tried to remove the call to `UnlockOSThread` but it didn't help.

I observed the following with the command:
```
strace -f -s 256 -e chdir,seccomp go test -v . -count 2 -v -run 'TestNotif$'
```

* First iteration of the test:
```
[pid 1901155] seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_NEW_LISTENER, {len=9, filter=0x1a70580}) = 6
[pid 1901155] chdir("/non-existent-path"    <unfinished ...>
[pid 1901155] <... chdir resumed>)      = -1 ENOENT (No such file or directory)
```
-> seccomp() returns the file descriptor correctly
-> chdir() returned ENOENT as expected

* Second iteration of the test:
```
[pid 1901157] seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=9, filter=0x7f6948001b50}) = 0
[pid 1901157] chdir("/non-existent-path") = -1 ENOSYS (Function not implemented)
```
-> The pid is different, so removing `UnlockOSThread` worked
-> seccomp() is not passed the flag SECCOMP_FILTER_FLAG_NEW_LISTENER as expected (why?)
-> seccomp() fails to return the file descriptor
-> chdir() does not return ENOENT but ENOSYS.

I am continuing to investigate...